src/AbstractCompositeContainer.php

```diff
@@ -62,11 +63,15 @@ protected function _createContainerId(BaseContainerInterface $container)
      *
      * @param string $id The ID of the service to retrieve.
      *
-     * @return mixed|null The service, if found; otherwise, null.
+     * @return mixed The service.
+     *
+     * @throws NotFoundException If none of the inner containers have a matching service.
      */
     protected function _getDelegated($id)
     {
-        $having = $this->_hasDelegated($id);
+        if (!($having = $this->_hasDelegated($id))) {
+            throw $this->_createNotFoundException(sprintf('Could not create service for ID "%1$s": no service defined', $id));
+        }
 
         return ($having)
```

Comment:
This check is now obsolete. I would also recommend inverting the `if` condition for cleaner code.

```php
if ($having = $this->_getDelegated($id)) {
    return $having->get($id)
}

throw $this->_createNotFoundException(sprintf('Could not create service for ID "%1$s": no service defined', $id));
```